web/interceptor/access_control.go

```diff
@@ -0,0 +1,181 @@
+		}
+		claims, err := tm.GetClaims(ctx)
+		if err != nil {
+			logger.Errorf("Access control: failed to get claims from request context: %v", err)
+			return handler(ctx, request)
+		}
+		for _, role := range roles {
+			switch role {
+			case none:
+				return handler(ctx, request)
+			case user:
+				if req.IDFor("user") == claims.UserID {
+					// Make sure the user is not updating own admin status.
+					if method == "UpdateUser" {
+						if req.(*qf.User).GetIsAdmin() != claims.Admin {
+							logger.Errorf("Access control: user %d attempted to change admin status from %v to %v",
+								claims.UserID, claims.Admin, req.(*qf.User).GetIsAdmin())
+						}
+					}
+					return handler(ctx, request)
+				}
+			case student:
+				// GetSubmissions is used to fetch individual and group submissions.
+				// For individual submissions needs an extra check for user ID in request.
+				if method == "GetSubmissions" && req.IDFor("group") == 0 {
+					if claims.UserID != req.IDFor("user") {
+						logger.Errorf("AccessControl: ID mismatch for %s in claims (%s) and request (%s)",
+							method, claims.UserID, req.IDFor("user"))
+						return nil, ErrAccessDenied
+					}
+				}
+				if claims.HasCourseStatus(req, qf.Enrollment_STUDENT) {
+					return handler(ctx, request)
+				}
+			case group:
+				// Request for CreateGroup will not have ID yet, need to check
+				// if the user is in the group (unless teacher).
+				if method == "CreateGroup" {
+					if claims.HasCourseStatus(req, qf.Enrollment_TEACHER) ||
+						req.(*qf.Group).Contains(&qf.User{ID: claims.UserID}) {
+						return handler(ctx, request)
+					} else {
+						logger.Errorf("AccessControl: user %d creates a new group while not teacher or group member")
+						return nil, ErrAccessDenied
+					}
+				}
+				groupID := req.IDFor("group")
+				for _, group := range claims.Groups {
+					if group == groupID {
+						return handler(ctx, request)
+					}
+				}
+			case teacher:
+				if method == "GetUserByCourse" {
+					if err := claims.IsCourseTeacher(tm.Database(), request.(*qf.CourseUserRequest)); err != nil {
+						logger.Errorf("AccessControl: %v", err)
+						return nil, ErrAccessDenied
+					}
+					return handler(ctx, request)
+				}
+				if claims.HasCourseStatus(req, qf.Enrollment_TEACHER) {
+					return handler(ctx, request)
+				}
+			case courseAdmin:
+				if claims.Admin {
+					if method == "GetUserByCourse" {
+						if err := claims.IsCourseTeacher(tm.Database(), request.(*qf.CourseUserRequest)); err != nil {
+							logger.Errorf("AccessControl: %v", err)
+							return nil, ErrAccessDenied
+						}
+						return handler(ctx, request)
+					}
+					if claims.HasCourseStatus(req, qf.Enrollment_TEACHER) {
+						return handler(ctx, request)
+					}
+				}
+			case admin:
+				if claims.Admin {
+					return handler(ctx, request)
+				}
+			}
+		}
+		logger.Errorf("Access denied (%s), required roles %v, user claims %v", method, access[method], claims)
```

Comment:
The output from this error is rather hard to parse:

```
Access denied (GetUsers), required roles [6], user claims {{ 1660727622  1660726722 QuickFeed 0 } 2 false map[] []}
```

We could implement the stringer interface on `Claims` and pass claims as `%s` with a sensible format.

Reply:
Fair point, added the `String()` method for user claims.